MCKIN-5257 Add data API to support native views and mobile downloads via Ooyala Mobile SDK

pomegranited opened the pull request:
Adds the `student_view_data` method, which returns the data necessary for streaming videos via the Ooyala Mobile SDK.

The `student_view_data` method also returns the `player_token` URL and `player_token_expires` timestamp. These values are not required for streaming playback from the Mobile SDK, but will be useful when the apps are updated to support video downloading (e.g [DownloadToOwnSampleApp](https://github.com/ooyala/ios-sample-apps/blob/2c7ce0584a7a88b34c08edc78daad29f070e7cec/PlaybackLab/DownloadToOwnSampleApp/DownloadToOwnSampleApp/Utils/BasicEmbedTokenGenerator.m)).

**JIRA tickets**: MCKIN-5257

**Partner information**: 3rd party-hosted open edX instance

**Testing instructions**:

1. Install this branch in your devstack:
    ```bash
    sudo su edxapp
    cd /edx/src/xblock-ooyala
    pip install -e .
    ```
1. Use `NO_PREREQ_INSTALL=1 paver run_all_servers --settings=devstack` to run both LMS and Studio in the same session.  The `run_all_servers` command is required to allow the LMS's Course Blocks API so that the cache can be cleared when Studio publishes a course.
1. In Studio, create a course .
1. Add `"ooyala-player"` to the advanced modules list
1. Add an Ooyala Player component to your course and configure it using your video Content ID, e.g.
    * Content ID: `5sMHA1YzE6KHI-dFSKgDz-pcMOx37_f9`
1. Verify that the video plays as expected.
1. Publish the unit.
1. Visit the [Course Blocks API](http://edx.readthedocs.io/projects/edx-platform-api/en/latest/courses/blocks.html#query-parameters) endpoint for your course, with the following parameters, e.g.,
    ```
    http://localhost:8000/api/courses/v1/blocks/?course_id=OpenCraft%2Fooyala%2F2017_1&all_blocks=true&depth=all&student_view_data=ooyala-player
    ```
    *Note:* If your LMS user doesn't have staff privileges, use `username=<lms_username>` instead of `all_blocks=true`.
    The output should show `student_view_data` for your `ooyala-player` block, e.g.
    ```js
    "i4x://OpenCraft/ooyala/ooyala-player/a16265917f5149daaa13c03c6fd8297d": {
            "type": "ooyala-player",
            ...
            "student_view_data": {
                "content_id": "5sMHA1YzE6KHI-dFSKgDz-pcMOx37_f9",
                "player_token_expires": null,
                "player_token": "",
                "partner_code": ""
            },
        },
    ```
1. Edit your Ooyala Player component to add your Partner Code, API Key and Secret.
   You will need an Ooyala Backlot account to perform this step.
    * Enable Player Token: True
    * Partner Code: See [docs](http://help.ooyala.com/video-platform/concepts/api_keys.html) for how to extract the pcode from your API key.
    * API Key: your API key
    * API Secret: your API secret
1. Publish the unit.
1. Refresh the Course Blocks API from step 8, and note the additional response data, e.g.,
    ```js
    "i4x://OpenCraft/ooyala/ooyala-player/a16265917f5149daaa13c03c6fd8297d": {
            "type": "ooyala-player",
            ...
            "student_view_data": {
                "content_id": "5sMHA1YzE6KHI-dFSKgDz-pcMOx37_f9",
                "player_token_expires": 1501812713,
                "player_token": "http://player.ooyala.com/sas/embed_token/53YWsyOszKOsfS1...",
                "partner_code": "53YWsyOszKOsfS1IvcQoKn8YhWYk"
            },
        },
    ```

To verify that this data API is sufficient for streaming videos from the Ooyala Mobile SDK, follow the Getting Started instructions on either the Ooyala [ios-sample-apps](https://github.com/ooyala/ios-sample-apps) or [android-sample-apps](https://github.com/ooyala/android-sample-apps) repos.

*Tip* I used `--depth 1` when cloning the sample app repos, because they're pretty large.

Both the iOS and Android sample repos contain a `BasicPlaybackSampleApp`, which displays a hardcoded list of videos.  The basic procedure verification procedure is:

1. Add an entry to that list using the `student_view_data` values.  Note that:
    * `embedCode` corresponds to `content_id`
    * `pcode` corresponds to `partner_code`
    * `domain` is where your XBlock runs from, e.g. `http://localhost:8000`
    * `player_token` is not required for video streaming, but will be used for video download.
1. Run the app.
1. Verify that your new video plays in the app as expected.
1. To verify that domain restrictions apply for video playback in the app, you'll need to use a video whose settings you can modify.
    * Login to the [Ooyala Backlot](https://backlot.ooyala.com/backlot/web#publish_tab).
    * Under Publish > Syndication Controls, tick the "Domain Controls" checkbox.
    * Add your devstack domain under the "Users cannot embed content" list of domains.
      Note that if you used "http://example.com:8001" in the video entry you added to the app, then you can block access from the app by using any of these variations: `http://example.com`, `http://example.com:8001`, `https://example.com`, `https://example.com:8000`
1. Try to watch your video again -- it should fail to play.


*iOS SDK*

Add a new [`PlayerSelectionOption`](https://github.com/ooyala/ios-sample-apps/blob/b8b9bba6bc86f4fd895efb2305d3b3a18378e056/BasicPlaybackSampleApp/BasicPlaybackSampleApp/Utils/PlayerSelectionOption.m#L14) to [BasicPlaybackListViewController.m](https://github.com/ooyala/ios-sample-apps/blob/b8b9bba6bc86f4fd895efb2305d3b3a18378e056/BasicPlaybackSampleApp/BasicPlaybackSampleApp/Lists/BasicPlaybackListViewController.m#L31-L35), e.g.,
```objectivec
    - (void)addAllBasicPlayerSelectionOptions {
      [self insertNewObject: [[PlayerSelectionOption alloc] initWithTitle:@"Kwatile"
                                                                embedCode:@"5sMHA1YzE6KHI-dFSKgDz-pcMOx37_f9"
                                                                    pcode:@"53YWsyOszKOsfS1IvcQoKn8YhWYk"
                                                                   domain:@"http://localhost:8001"
                                                           viewController: [BasicSimplePlayerViewController class]]];
    ...
```

*Android SDK*

Add a new add a new [`PlayerSelectionOption`](https://github.com/ooyala/android-sample-apps/blob/640e5e400eb6af519c95b38cb544d9d1de5ed416/BasicPlaybackSampleApp/app/src/main/java/com/ooyala/sample/utils/PlayerSelectionOption.java#L9) to  [BasicPlaybackListActivity.java](https://github.com/ooyala/android-sample-apps/blob/stable/BasicPlaybackSampleApp/app/src/main/java/com/ooyala/sample/lists/BasicPlaybackListActivity.java#L37), e.g.,
```java
  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setTitle(getName());

    selectionMap = new LinkedHashMap<String, PlayerSelectionOption>();
    //Populate the embed map
    selectionMap.put( "Kwatile", new PlayerSelectionOption("5sMHA1YzE6KHI-dFSKgDz-pcMOx37_f9", "53YWsyOszKOsfS1IvcQoKn8YhWYk", "http://localhost:8001", BasicPlaybackVideoPlayerActivity.class) );
    ...
```

**Author Notes & Concerns**

Since rebasing on `player_v4_with_translations`, the unit tests are failing for two tests unrelated to this change. CC @naeem91 
```
======================================================================
ERROR: tests.test_ooyala_player.test_studio_submit_json_handler_valid_input
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/edx/src/venvs/xblock-ooyala/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/edx/src/xblock-ooyala/tests/test_ooyala_player.py", line 127, in test_studio_submit_json_handler_valid_input
    result = player.studio_submit(_MockRequest(request_data))
  File "/edx/src/venvs/xblock-ooyala/local/lib/python2.7/site-packages/xblock/mixins.py", line 68, in wrapper
    response = func(self, request_json, suffix)
  File "/edx/src/xblock-ooyala/ooyala_player/ooyala_player.py", line 439, in studio_submit
    self.disable_cc_and_translations = submissions['cc_disable']
KeyError: 'cc_disable'

======================================================================
ERROR: tests.test_ooyala_player.test_studio_submit_json_handler_another_valid_input
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/edx/src/venvs/xblock-ooyala/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/edx/src/xblock-ooyala/tests/test_ooyala_player.py", line 153, in test_studio_submit_json_handler_another_valid_input
    result = player.studio_submit(_MockRequest(request_data))
  File "/edx/src/venvs/xblock-ooyala/local/lib/python2.7/site-packages/xblock/mixins.py", line 68, in wrapper
    response = func(self, request_json, suffix)
  File "/edx/src/xblock-ooyala/ooyala_player/ooyala_player.py", line 439, in studio_submit
    self.disable_cc_and_translations = submissions['cc_disable']
KeyError: 'cc_disable'
```

**Reviewers**
- [ ] @haikuginger 
- [ ] edX reviewer[s] TBD

naeem91 commented:
@pomegranited please note; `master` does not contain the latest code yet. Ooyala latest release is set to tag which points to latest commit in https://github.com/edx-solutions/xblock-ooyala/tree/player_v4_with_translations

pomegranited commented:
@naeem91  Thank you for the info!  Would you like me to target that branch with this PR instead of master?

naeem91 commented:
Yes. You may point and test with that branch and I'll try to make master reflect the latest code shortly

haikuginger commented on tests/utils.py:
```diff
@@ -0,0 +1,31 @@
+from xblock.test.tools import TestRuntime
+
+
+class MockNow:
```
Since we're running in Python 2, please inherit from `object`.

bradenmacdonald commented:
@pomegranited I'm wondering if we can avoid the need for the `player_data` handler. The only reason for it is to get the 'domain' in a nice way, but it will require an additional API call every time this API is used from the mobile apps, which will slow things down slightly, and performance is at a premium for this app.

So I'd like to suggest that for now, we don't provide 'domain' from the XBlock API, and we just assume that the mobile app frontend code will know what the domain is and fill in that setting. After all, the app is calling this API in the first place, so the app already does have the domain coded as a setting (which it uses to determine the full API URL).

If for some reason we had to keep `player_data` in, should it be a GET request than a POST request?

pomegranited commented:
@haikuginger  Thanks for your review! I've addressed your nit, and have removed the `player_data` endpoint as @bradenmacdonald  suggested - he's totally right; if we've got the `player_data` endpoint URL, then we've got the `domain`, so there's no need to actually call it.  Have updated the PR description to illustrate how to test using the Course Blocks API instead of the new endpoint.

I've also rebased d9208ac on [player_v4_with_translations](https://github.com/edx-solutions/xblock-ooyala/tree/player_v4_with_translations), and fixed the conflicts.

Sorry I didn't get this rebase done in time for you to review the whole thing at once yesterday.  Do you mind to do another quick check?

CC @naeem91

haikuginger commented:
:+1:, again!

- [x] I tested this: followed the instructions and got the expected results.
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

bradenmacdonald commented:
@pomegranited @haikuginger Does this actually include documentation? I don't see any student_view_data examples in the README or docstrings.
